Add support for BME680 SPI Interface

leonardp opened the pull request:
This Pull request adds support for the BME680 SPI Interface.
It is mostly analogous to the BME280.

Signed-off-by: Leonard Pollak leonardp@tr-host.de

teburd reviewed (dismissed):
As mentioned the compile ifdef for setting the mem page when using SPI seems an issue to me as well, everything else looks great

mbolivar-nordic commented on drivers/sensor/bme680/bme680.c:
```diff
@@ -6,37 +6,86 @@
+				  uint8_t start, uint8_t *buf, int size)
 {
 	const struct bme680_config *config = dev->config;
+#if BME680_BUS_SPI
```
Why not fold this into the individual bus_io->read routines as needed instead?

leonardp replied:
This is roughly what I ended up doing.
I made a separate function that does the mem page switching and I am also calling it from the  bus_io->write routine.

leonardp commented:
Sorry for shelving this for so long without notice.
Thanks for the reviews so far.
I will (most probably) be able to get this done by the end of next week.

leonardp commented:
I've updated the PR and tested it with I2C and SPI instance enabled simultaneously.

mbolivar-nordic commented on drivers/sensor/bme680/bme680.c:
```diff
@@ -342,11 +350,28 @@ static int bme680_read_compensation(const struct device *dev)
+	}
+
+	/*
+	 * When using both (I2C and SPI) sensors, this read operation will also
+	 * be executed by the I2C device.
```
Can you clarify why it needs to be in the ifdef, then, if it's safe to do for both bus types?

leonardp replied:
It is safe for both bus types. The idea was to save a little bit of code/time on case only i2c is enabled.
I'll edit the comment to explain this more clearly, ok?

mbolivar-nordic commented on drivers/sensor/bme680/bme680.h:
```diff
@@ -29,19 +74,22 @@
 #define BME680_REG_CTRL_HUM             0x72
 #define BME680_REG_CTRL_MEAS            0x74
 #define BME680_REG_CONFIG               0x75
-#define BME680_REG_MEM_PAGE             0xf3
+#define BME680_REG_MEM_PAGE             0x73
```
There are several changes in this commit that are unrelated to adding support for the SPI interface. For the most part, they are minor enough that I don't want to quibble over them. But changes like this really shouldn't be squashed into a commit that claims to be only adding support for a new bus type. Please split this into its own commit with a separate justification in the commit log, so it can be reviewed separately. If the register value is simply wrong, it's a separate bug fix patch.

(Ideally, if you can take the time to split other unrelated changes (like moving code around, adding log messages, etc.) into a separate 'prep work' commit as well, that would make this whole patch easier to review, but as I said, I otherwise wouldn't have quibbled this time.)

leonardp replied:
The register value is not *really* wrong. Confusingly it just had the write bit set already.
Regardless, your criticism is well justified and I will take the time to split up the unrelated changes into their own commits, get rid of the log messages and add constants for the SPI read/write masks :+1:

MaureenHelm replied:
Thank you for doing this

mbolivar-nordic reviewed (changes requested):
Sorry for the long silence; I'm just catching up on older reviews now that the merge window is about to reopen for Zephyr v3.1. I noticed a few minor things, but otherwise this looks quite good; nice work.

Commit:
drivers: sensor: bme680: fix constants

This fixes the constant for the mem page and replaces a
magic number with the already defined `BME680_LEN_COEFF2` constant.

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

Commit:
drivers: sensor: bme680: prep work

Consolidate the initialization routines and change the include guard to
conform with the coding guidelines as a preparation for the following
commits which add support for the SPI interface.

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

Commit:
drivers: sensor: bme680: Add SPI interface

This enables the SPI interface for the BME680 sensor driver.

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

Commit:
tests: drivers: Add bme680-spi to build_all

Add BME680 (SPI) sensor driver to build_all tests.

Signed-off-by: Leonard Pollak <leonardp@tr-host.de>

leonardp commented:
I split the pull request into 4 commits as suggested by @mbolivar-nordic.
I hope everything is in order now.
Thanks for all the reviews so far!

mbolivar-nordic commented on drivers/sensor/bme680/bme680.c:
```diff
@@ -346,13 +352,25 @@ static int bme680_read_compensation(const struct device *dev)
+	}
+
+	/*
+	 * In case only I2C is used this saves a bit of code/time.
+	 * When using both (I2C and SPI) sensors it is safe to execute
+	 * this for the I2C device.
+	 */
+#if BME680_BUS_SPI
```
Thanks for explaining. In that case I think it would be a little better to write it something like this:

```c
static inline bool bme680_is_on_spi(struct device *dev)
{
	const struct bme680_config *config = dev->config;

	return config->bus_io == &bme680_bus_io_spi;
}

/* ... */

#if BME680_BUS_SPI
	if (bme680_is_on_spi(dev)) {
		err = ...;
		if (err < 0) {
			return err;
		}
	}
#endif
```

IMO it is a bit better to have the I2C devices initialize the same way always, regardless of whether a SPI device happens to be in the same system. It may ease debugging in some cases and it definitely enhances consistency.

But it's not a blocking request; just please consider it.

leonardp replied:
Thanks for taking another look at it and giving some more insight on that matter.

To have it consistent across all possible configurations wouldn't it make sense to 
remove the `#if BME680_BUS_SPI` directive aswell?

I have prepared the change as suggested by you (without the `#if`). If that makes sense to you I can push them.

mbolivar-nordic replied:
That's a fair point -- if it's safe to do regardless of bus type, might as well be consistent on all of them. Given the ifdef, I was assuming there was a reason to prefer not to do it on I2C, but if there isn't, then I'm fine either way.

leonardp replied:
Well, it was probably me just over-complicating things -.-;
I pushed the changes, this way it is definitely easier on the eyes.

leonardp replied:
...disregard that.. the directives are needed or else it won't build for I2C, obviously..
Sorry for the noise.

mbolivar-nordic reviewed (dismissed):
Thank you, that was indeed easier to review

leonardp commented:
@teburd would you take another look?

MaureenHelm commented:
I think this PR got lost in the shuffle while the merge window was closed for the 3.0 release.

@leonardp will you please rebase and resolve the merge conflict?

carlescufi commented:
@MaureenHelm @mbolivar-nordic I rebased this, please re-ack.

Senso4sLab commented:
I tested bme680 sample with SPI and I got the following error <err> bme680: Bad BME680 chip id: 0x0. I used the following prj.conf settings

```
&spi0 {
	compatible = "nordic,nrf-spi";	
	status = "okay";
	pinctrl-0 = <&spi0_default>;
	pinctrl-1 = <&spi0_sleep>;
	pinctrl-names = "default", "sleep";
    cs-gpios = <&gpio0 13 GPIO_ACTIVE_LOW>,
               <&gpio1 6 GPIO_ACTIVE_LOW>;              
    
    bme6800:bme680_0@0 {
		compatible = "bosch,bme680";
        status = "okay";
		reg = <0>;
        label="bme680_1";
		spi-max-frequency = <1000000>;        
	};
    bme6801:bme680_1@1 {
		compatible = "bosch,bme680";
        status = "okay";
		reg = <1>;
        label="bme680_2";
		spi-max-frequency = <1000000>; 
	};
     
};
```

I will appreciate for help!
